### PR TITLE
`Required` rule: remove `SkipOnEmptyInterface` and add `emptyCallback` option

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #336

### CODE REVIEW

**Review by samdark — APPROVED** [review]
👍 https://github.com/yiisoft/validator/blob/master/README.md#skipping-empty-values should be adjusted though.

File: src/Rule/Required.php [hunk 1]
```diff
@@ -6,39 +6,46 @@
...
 {
-    use SkipOnEmptyTrait;
     use SkipOnErrorTrait;
     use WhenTrait;
 
+    /**
+     * @var callable
+     * @psalm-var EmptyCallback
+     */
+    private $emptyCallback;
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Rule/Required.php:32]: The `callable` type is not supported. See [RFC](https://wiki.php.net/rfc/typed_properties_v2#supported_types).

[the remark this answers is not in the document]
Comment (samdark) [reply src/Rule/Required.php:32]: You're right. OK :)

File: tests/ValidatorTest.php [hunk 2]
```diff
@@ -862,39 +862,39 @@ public function initSkipOnEmptyDataProvider(): array
             'null' => [
                 null,
                 new class () {
-                    #[Required]
```
Comment (arogachev) [inline tests/ValidatorTest.php:865]: We can leave / add one test with required to make sure this check works:

https://github.com/yiisoft/validator/blob/b77dedd843d0157e8be8879dab811a68d17bc7a5/src/Validator.php#L198-L200

Comment (vjik) [reply tests/ValidatorTest.php:865]: Agree, but PR about another. Suggest make this later, when we will be increase psalm to 1 and kill mutants.

Comment (arogachev) [reply tests/ValidatorTest.php:865]: OK. Let's do it in a separate PR.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,4 +2,4 @@
 
 ## 1.0.0 under development
 
-- Initial release.
+- Enh: Add composer require checker into CI
```
Comment (rustamwin) [inline CHANGELOG.md:5]: ?

Comment (arogachev) [reply CHANGELOG.md]: The diff was shown wrong. Now it's OK.

File: README.md [hunk 4]
```diff
@@ -142,7 +142,10 @@ new Number(asInteger: true, max: 100, skipOnError: true)
 #### Skipping empty values
 
 By default, missing and empty values are validated (if the value is missing, it's considered `null`). That is
-undesirable if you need to allow not specifying a field. To change this behavior, use `skipOnEmpty: true`:
+undesirable if you need to allow not specifying a field. To change this behavior, use `skipOnEmpty: true`.
```
Comment (samdark) [inline README.md:145]:
```suggestion
undesirable if you need a field to be optional. To change this behavior, use `skipOnEmpty: true`.
```

File: README.md [hunk 5]
```diff
@@ -142,7 +142,10 @@ new Number(asInteger: true, max: 100, skipOnError: true)
...
+
+Note that not all rules have this option, but only those that implement `Yiisoft\Validator\SkipOnEmptyInterface`.
+For example, `Required` rule doesn't. For more details about its customization see "Requiring values" section.
```
Comment (samdark) [inline README.md:148]:
```suggestion
Note that not every rule has this option, but only the ones that implement `Yiisoft\Validator\SkipOnEmptyInterface`.
For example, `Required` rule doesn't. For more details see "Requiring values" section.
```

File: README.md [hunk 6]
```diff
@@ -700,6 +703,19 @@ final class Post
...
+Use `Yiisoft\Validator\Rule\Required` rule to make sure value is presented. What values exactly are considered empty can
+be customized via `$emptyCallback` option. Normalization is not performed here, so only a callable or special class is 
+needed. For more details see "Skipping empty values" section.
+
+```php
+use Yiisoft\Validator\Rule\Required;
+use Yiisoft\Validator\SkipOnEmptyCallback\SkipOnNull;
+
+$rules = [new Required(emptyCallback: new SkipOnNull())];
+```
```
Comment (samdark) [inline README.md:717]:
```suggestion
Use `Yiisoft\Validator\Rule\Required` rule to make sure a value is present. What values are considered empty can
be customized via `$emptyCallback` option. Normalization is not performed, so only a callable or special class is 
needed. For more details see "Skipping empty values" section.

```php
use Yiisoft\Validator\Rule\Required;
use Yiisoft\Validator\SkipOnEmptyCallback\SkipOnNull;

$rules = [new Required(emptyCallback: new SkipOnNull())];
```
```

**Pull request merged**